File: pysrc/Random.cpp
/* -*- c++ -*-
 * Copyright (c) 2012-2023 by the GalSim developers team on GitHub
 * https://github.com/GalSim-developers
 *
 * This file is part of GalSim: The modular galaxy image simulation toolkit.
 * https://github.com/GalSim-developers/GalSim
 *
 * GalSim is free software: redistribution and use in source and binary forms,
 * with or without modification, are permitted provided that the following
 * conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions, and the disclaimer given in the accompanying LICENSE
 *    file.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions, and the disclaimer given in the documentation
 *    and/or other materials provided with the distribution.
 */

#include "PyBind11Helper.h"
#include "Random.h"
#include <numpy/random/bitgen.h>  // For bitgen_t struct

// For reference:
/*
typedef struct bitgen {
  void *state;
  uint64_t (*next_uint64)(void *st);
  uint32_t (*next_uint32)(void *st);
  double (*next_double)(void *st);
  uint64_t (*next_raw)(void *st);
} bitgen_t;
*/

namespace galsim {

    uint32_t next_uint32(void* st)
    {
        BaseDeviate* rng = static_cast<BaseDeviate*>(st);
        return (uint32_t) rng->raw();
    }

    uint64_t next_uint64(void* st)
    {
        uint32_t i1 = next_uint32(st);
        uint32_t i2 = next_uint32(st);
        return ((uint64_t) i1 << 32) + i2;
    }

    double next_double(void* st)
    {
        return next_uint32(st) / ((double) std::numeric_limits<uint32_t>::max()+1.);
    }

    void SetupBitGen(BaseDeviate* rng, py::capsule capsule)
    {
        bitgen_t* bg(capsule);
        bg->state = rng;
        bg->next_uint64 = next_uint64;
        bg->next_uint32 = next_uint32;
        bg->next_double = next_double;
        bg->next_raw = next_uint64;
    };

    void Generate(BaseDeviate& rng, size_t N, size_t idata)
    {
        double* data = reinterpret_cast<double*>(idata);
        rng.generate(N, data);
    }

    void AddGenerate(BaseDeviate& rng, size_t N, size_t idata)
    {
        double* data = reinterpret_cast<double*>(idata);
        rng.addGenerate(N, data);
    }

    void GenerateFromVariance(GaussianDeviate& rng, size_t N, size_t idata)
    {
        double* data = reinterpret_cast<double*>(idata);
        rng.generateFromVariance(N, data);
    }

    void GenerateFromExpectation(PoissonDeviate& rng, size_t N, size_t idata)
    {
        double* data = reinterpret_cast<double*>(idata);
        rng.generateFromExpectation(N, data);
    }

    void pyExportRandom(py::module& _galsim)
    {
        py::class_<BaseDeviate> (_galsim, "BaseDeviateImpl")
            .def(py::init<long>())
            .def(py::init<const BaseDeviate&>())
            .def(py::init<const char*>())
            .def("duplicate", &BaseDeviate::duplicate)
            .def("seed", (void (BaseDeviate::*) (long) )&BaseDeviate::seed)
            .def("reset", (void (BaseDeviate::*) (const BaseDeviate&) )&BaseDeviate::reset)
            .def("clearCache", &BaseDeviate::clearCache)
            .def("serialize", &BaseDeviate::serialize)
            .def("discard", &BaseDeviate::discard)
            .def("raw", &BaseDeviate::raw)
            .def("generate", &Generate)
            .def("add_generate", &AddGenerate)
            .def("setup_bitgen", &SetupBitGen)
            ;

        py::class_<UniformDeviate, BaseDeviate>(
            _galsim, "UniformDeviateImpl")
            .def(py::init<const BaseDeviate&>())
            .def("duplicate", &UniformDeviate::duplicate)
            .def("generate1", &UniformDeviate::generate1);

        py::class_<GaussianDeviate, BaseDeviate>(
            _galsim, "GaussianDeviateImpl")
            .def(py::init<const BaseDeviate&, double, double>())
            .def("duplicate", &GaussianDeviate::duplicate)
            .def("generate1", &GaussianDeviate::generate1)
            .def("generate_from_variance", &GenerateFromVariance);

        py::class_<BinomialDeviate, BaseDeviate>(
            _galsim, "BinomialDeviateImpl")
            .def(py::init<const BaseDeviate&, int, double>())
            .def("duplicate", &BinomialDeviate::duplicate)
            .def("generate1", &BinomialDeviate::generate1);

        py::class_<PoissonDeviate, BaseDeviate>(
            _galsim, "PoissonDeviateImpl")
            .def(py::init<const BaseDeviate&, double>())
            .def("duplicate", &PoissonDeviate::duplicate)
            .def("generate1", &PoissonDeviate::generate1)
            .def("generate_from_expectation", &GenerateFromExpectation);

        py::class_<WeibullDeviate, BaseDeviate>(
            _galsim, "WeibullDeviateImpl")
            .def(py::init<const BaseDeviate&, double, double>())
            .def("duplicate", &WeibullDeviate::duplicate)
            .def("generate1", &WeibullDeviate::generate1);

        py::class_<GammaDeviate, BaseDeviate>(
            _galsim, "GammaDeviateImpl")
            .def(py::init<const BaseDeviate&, double, double>())
            .def("duplicate", &GammaDeviate::duplicate)
            .def("generate1", &GammaDeviate::generate1);

        py::class_<Chi2Deviate, BaseDeviate>(
            _galsim, "Chi2DeviateImpl")
            .def(py::init<const BaseDeviate&, double>())
            .def("duplicate", &Chi2Deviate::duplicate)
            .def("generate1", &Chi2Deviate::generate1);
    }

} // namespace galsim

File: galsim/deprecated/random.py
# Copyright (c) 2012-2017 by the GalSim developers team on GitHub
# https://github.com/GalSim-developers
#
# This file is part of GalSim: The modular galaxy image simulation toolkit.
# https://github.com/GalSim-developers/GalSim
#
# GalSim is free software: redistribution and use in source and binary forms,
# with or without modification, are permitted provided that the following
# conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice, this
#    list of conditions, and the disclaimer given in the accompanying LICENSE
#    file.
# 2. Redistributions in binary form must reproduce the above copyright notice,
#    this list of conditions, and the disclaimer given in the documentation
#    and/or other materials provided with the distribution.
#


import galsim
from galsim.deprecated import depr

def GaussianDeviate_setMean(self, mean):
    """Deprecated method to set the mean.
    """
    depr('setMean', 1.1, 'rng = galsim.GaussianDeviate(rng, mean=mean, sigma=rng.sigma)')
    self._setMean(mean)

def GaussianDeviate_setSigma(self, sigma):
    """Deprecated method to set sigma.
    """
    depr('setSigma', 1.1, 'rng = galsim.GaussianDeviate(rng, mean=rng.mean, sigma=sigma)')
    self._setSigma(sigma)

def GaussianDeviate_getMean(self):
    depr('gd.getMean', 1.5, 'gd.mean')
    return self.mean

def GaussianDeviate_getSigma(self):
    depr('gd.getSigma', 1.5, 'gd.sigma')
    return self.sigma

galsim._galsim.GaussianDeviate.setMean = GaussianDeviate_setMean
galsim._galsim.GaussianDeviate.setSigma = GaussianDeviate_setSigma
galsim._galsim.GaussianDeviate.getMean = GaussianDeviate_getMean
galsim._galsim.GaussianDeviate.getSigma = GaussianDeviate_getSigma

def BinomialDeviate_setN(self, N):
    """Deprecated method to set N.
    """
    depr('setN', 1.1, 'rng = galsim.BinomialDeviate(rng, N=N, p=rng.p)')
    self._setN(N)

def BinomialDeviate_setP(self, p):
    """Deprecated method to set p.
    """
    depr('setP', 1.1, 'rng = galsim.BinomialDeviate(rng, N=rng.N, p=p)')
    self._setP(p)

def BinomialDeviate_getN(self):
    depr('bd.getN', 1.5, 'bd.n')
    return self.n

def BinomialDeviate_getP(self):
    depr('bd.getP', 1.5, 'bd.p')
    return self.p

galsim._galsim.BinomialDeviate.setN = BinomialDeviate_setN
galsim._galsim.BinomialDeviate.setP = BinomialDeviate_setP
galsim._galsim.BinomialDeviate.getN = BinomialDeviate_getN
galsim._galsim.BinomialDeviate.getP = BinomialDeviate_getP

def PoissonDeviate_setMean(self, mean):
    """Deprecated method to set the mean.
    """
    depr('setMean', 1.1, 'rng = galsim.PoissonDeviate(rng, mean=mean)')
    self._setMean(mean)

def PoissonDeviate_getMean(self):
    depr('pd.getMean', 1.5, 'pd.mean')
    return self.mean

galsim._galsim.PoissonDeviate.setMean = PoissonDeviate_setMean
galsim._galsim.PoissonDeviate.getMean = PoissonDeviate_getMean

def WeibullDeviate_setA(self, a):
    """Deprecated method to set a.
    """
    depr('setA', 1.1, 'rng = galsim.WeibullDeviate(rng, a=a, b=rng.b)')
    self._setA(a)

def WeibullDeviate_setB(self, b):
    """Deprecated method to set b.
    """
    depr('setB', 1.1, 'rng = galsim.WeibullDeviate(rng, a=rng.a, b=b)')
    self._setB(b)

def WeibullDeviate_getA(self):
    depr('wd.getA', 1.5, 'wd.a')
    return self.a

def WeibullDeviate_getB(self):
    depr('wd.getB', 1.5, 'wd.b')
    return self.b

galsim._galsim.WeibullDeviate.setA = WeibullDeviate_setA
galsim._galsim.WeibullDeviate.setB = WeibullDeviate_setB
galsim._galsim.WeibullDeviate.getA = WeibullDeviate_getA
galsim._galsim.WeibullDeviate.getB = WeibullDeviate_getB

def GammaDeviate_setK(self, k):
    """Deprecated method to set k.
    """
    depr('setK', 1.1, 'rng = galsim.GammaDeviate(rng, k=k, theta=rng.theta)')
    self._setK(k)

def GammaDeviate_setTheta(self, theta):
    """Deprecated method to set theta.
    """
    depr('setTheta', 1.1, 'rng = galsim.GammaDeviate(rng, k=rng.k, theta=theta)')
    self._setTheta(theta)

def GammaDeviate_getK(self):
    depr('gd.getK', 1.5, 'gd.k')
    return self.k

def GammaDeviate_getTheta(self):
    depr('gd.getTheta', 1.5, 'gd.theta')
    return self.theta

galsim._galsim.GammaDeviate.setK = GammaDeviate_setK
galsim._galsim.GammaDeviate.setTheta = GammaDeviate_setTheta
galsim._galsim.GammaDeviate.getK = GammaDeviate_getK
galsim._galsim.GammaDeviate.getTheta = GammaDeviate_getTheta

def Chi2Deviate_setN(self, n):
    """Deprecated method to set n.
    """
    depr('setN', 1.1, 'rng = galsim.Chi2Deviate(rng, n=n)')
    self._setN(n)

def Chi2Deviate_getN(self):
    depr('cd.getN', 1.5, 'cd.n')
    return self.n

galsim._galsim.Chi2Deviate.setN = Chi2Deviate_setN
galsim._galsim.Chi2Deviate.getN = Chi2Deviate_getN

